collector.go

```diff
@@ -1,53 +1,87 @@
 package sync
 
 import (
+	"context"
+	"errors"
+	"iter"
 	"sync"
 )
 
-type ProviderFunc[T any] func() T
-
-// Collector describes an executor that runs providers in parallel and returns the results from all providers
-// after they have completed by using the Collect function
-type Collector[T any] interface {
-	// Provide is used to add providers to this collector
-	Provide(provider ProviderFunc[T])
+// Collect iterates over the provided values, executing the processor in parallel to map each incoming value to a result.
+// The collector is used to apply the results, with an exclusive lock; collector will never execute in parallel.
+// All errors returned from processor functions will be joined with errors.Join as the returned error.
+func Collect[From, To any](ctx *context.Context, executorName string, values iter.Seq[From], collector func(From, To), processor func(From) (To, error)) error {
```

Comment:
this is a design nit: there are three things that happen in this rough order:
- iterate values
- process values
- collect values

In terms of the first element looked at (regardless of concurrency) this is the order it follows. I feel like these things should be passed to collect in the same order they will be used:

```suggestion
func Collect[From, To any](ctx *context.Context, executorName string, values iter.Seq[From], processor func(From) (To, error), collector func(From, To)) error {
```

I also think that `values` should really be called `iterator` here.

Reply:
nit on naming: should `collect` really be `assign`?

Reply:
The problem with changing the order is that the big function ends up being in the middle of the arguments rather than last. So, a simple pattern to follow of defining the function inline, e.g. [here](https://github.com/anchore/syft/pull/3636/files#diff-3d8bf2b1b904de169e43c2a64173c1e628d5f928820db80a7932be9d9e870905R54) ends up "feeling" a bit off to me and requires putting map/slice at the end in the specialized versions for consistency.

Naming: I don't think "assign" is really right here, either. It's essentially a "parallel reduce" function, these often refer to this as the accumulator.

I've made these changes, regardless.

Reply:
I'm ok if you wanted to leave it as is too, you don't have to incorporate these changes